### PR TITLE
Corrects validation of secrets in extension installation

### PR DESCRIPTION
**Pull request by philnash**
### Description

[Discovered while trying to install the Twilio send-messages extension](https://github.com/twilio-labs/twilio-firebase-extensions/issues/13), even with a correctly formatted secret, it never passed the validation rules (though it was added to secrets manager).

The issue was that the methods `promptCreateSecret` and `promptReconfigureSecret` were returning the result of `addNewSecretVersion` which looked like this: `projects/${version.secret.projectId}/secrets/${version.secret.name}/versions/${version.versionId}`. This string then failed to pass the secret validation.

This pull request returns the original string value entered as the secret allowing validation to be performed correctly. Except in the case of choosing not to reconfigure the secret, in which case there is no new secret and the existing secret is deemed valid.

---

This update does not change the fact that an invalid secret will be stored in Secrets Manager before the validation check is done. A bigger refactor of this code could rearrange that order to save on API requests when a secret is invalid. That refactor would also benefit from test cases for select, multiselect and secret types on top of the existing string type.

### CODE REVIEW

**Comment by i14h** [comment]
@pavelgj since Joe is out today

**Comment by philnash** [comment]
@pavelgj I've updated this so that `promptCreateSecret` and `promptReconfigureSecret` handle the validation of the response within themselves, which simplifies `askForParam`. Both `promptCreateSecret` and `promptReconfigureSecret` both respond with the secret ID that is created (which I think would have broken things before).

My only question is what is supposed to happen when you are reconfiguring and you choose to leave a secret as it is? You don't seem to have access to the existing secret ID and returning an empty string feels wrong (though is what I have implemented for now).

File: src/test/extensions/askUserForParam.spec.ts [hunk 1]
```diff
@@ -198,7 +200,7 @@ describe("askUserForParam", () => {
       expect(res).to.equal("");
     });
   });
-  describe("askForParam", () => {
+  describe("askForParam with string param", () => {
```
Comment (pavelgj) [inline src/test/extensions/askUserForParam.spec.ts:203]: undo?

File: src/extensions/askUserForParam.ts [hunk 2]
```diff
@@ -157,23 +160,33 @@ async function promptReconfigureSecret(
...
+        } else {
+          return promptReconfigureSecret(projectId, instanceId, paramSpec);
+        }
+      } else {
+        return "";
       }
-      return addNewSecretVersion(projectId, instanceId, secret, paramSpec, secretValue);
     case SecretUpdateAction.LEAVE:
     default:
-      return paramSpec.default || "";
```
Comment (pavelgj) [inline src/extensions/askUserForParam.ts:176]: I believe this should be `paramSpec.default || ""` -- `default` contains previous value, which we want in the `LEAVE` case.

**Review by pavelgj — DISMISSED** [review]

File: src/test/extensions/askUserForParam.spec.ts (hunk 1 above)
Comment (philnash) [reply src/test/extensions/askUserForParam.spec.ts:203]: This was to contrast with the [new test below "askForParam with secret param"](https://github.com/firebase/firebase-tools/pull/3901/files#diff-3e09552fce53dd0c656d89aaaea5adcffb1e8b33316dbfd5332afbf35fce1a9bR223).

File: src/extensions/askUserForParam.ts (hunk 2 above)
Comment (philnash) [reply src/extensions/askUserForParam.ts:176]: Thanks! Updated.

File: src/extensions/askUserForParam.ts [hunk 3]
```diff
@@ -113,11 +113,15 @@ export async function askForParam(
             "You may select multiple options.",
           choices: convertExtensionOptionToLabeledList(paramSpec.options as ParamOption[]),
         });
+        valid = checkResponse(response, paramSpec);
         break;
       case ParamType.SECRET:
-        response = reconfiguring
-          ? await promptReconfigureSecret(projectId, instanceId, paramSpec)
-          : await promptCreateSecret(projectId, instanceId, paramSpec);
+        if (reconfiguring) {
```
Comment (joehan) [inline src/extensions/askUserForParam.ts:119]: Nit: Prefer this as a ternary, like it was before.

**Review by joehan — APPROVED** [review]
One nite, but code LGTM. Thanks for fixing this!

**Comment by joehan** [comment]
Could you also add an entry to CHANGELOG.md describing this fix? Something like:
- Fixes issue where secrets for Firebase Extensions would be created before validation.

**Comment by philnash** [comment]
Updated those two bits @joehan.

**Comment by philnash** [comment]
@joehan Is this good to be merged now? Anything else you need from me?

**Comment by joehan** [comment]
@philnash Nope, we had a code freeze on for a while (and then this slipped past me for a few days 😬). This looks good to go - thanks for your contribution!

**Pull request merged**